CAToolsetPerfDataCollector/PerfDataCollector.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <iostream>
+#include <CppBuildInsights.hpp>
+#include <map>
+#include <unordered_map>
+#include <compare>
+
+#include <comutil.h>
+
+using namespace Microsoft::Cpp::BuildInsights;
+using namespace Activities;
+using namespace SimpleEvents;
+using namespace std::chrono;
+
+struct PathId
+{
+    int id;
+    const std::wstring* path;
+
+    constexpr auto operator<=>(const PathId& b) const { return *path <=> *b.path; };
+};
+
+struct PerfData
+{
+    // Notes on FrontEndPass, CodeAnalysisPass, and BackEndPass:
+    //  - FrontEndPass can be omitted if /analyze:only is used. In that case, the CodeAnalysisPass is the only pass.
+    //  - CodeAnalysisPass is a special FrontEndPass for Code Analysis when /analyze option is used, and includes actual analysis of code.
+    //  - BackEndPass can be omitted if /c and/or /analyze:only is used.
+    microseconds FEPass;                    // FrontEnd pass
+    microseconds BEPass;                    // BackEnd pass
+    microseconds CAPass;                    // CodeAnalysis pass
+    microseconds ASTCreation;               // AST creation for Code Analysis
+    microseconds ASTClients;                // COM clients for Code Analysis, including PREfast
+    microseconds CAFunction;                // PREfast's Code Analysis of functions, including analysis of function by all of its plug-ins
+    microseconds FPAFunction;               // PREfast's path-sensitive Code Analysis of functions (FPA = Function Path Analysis)
+    microseconds EspXCfgBuild;              // CFG Building for functions to be used by EspXEngine and its extensions
+    microseconds EspXAllChecks;             // Analysis of functions by EspXEngine and its extensions
+    microseconds EspXPathSensitiveChecks;   // Path-sensitive analysis of functions by EspXEngine and its extensions
+
+    // This header is used to print the performance data to the console.
+    // It should be updated if the PerfData struct is updated.
+    // Note that this header does not include the file path, which should be printed separately if needed.
+    static constexpr std::wstring_view PerfDataHeader =
+        L"FrontEnd Pass, BackEnd Pass, CodeAnalysis Pass, "
+        L"AST Creation, AST Clients, "                                      // Children of CodeAnalysis Pass
+        L"Function Analysis, "                                              // Child of AST Clients
+        L"FPA Function Analysis, EspX CFG Build, EspX Function Analysis, "  // Children of Function Analysis
+        L"EspX Path-sensitive Analysis";                                    // Child of EspX Function Analysis
+};
+
+class PerfDataCollector : public IAnalyzer
+{
+private:
+    std::unordered_map<std::wstring, PathId> filePaths{};   // List of file paths
+    std::map<PathId, PerfData> perfDataPerTu{};                  // Performance data per TU
```

Comment:
Why do we need the PathId? Could we use the paths directly?

Reply:
This is to reduce size and lookup. Comparison of integral value vs full string comparison.

Reply:
I will leave it as is, just in case we will have records per-function in the future. In that case, we will want to not repeat full paths for each per-function records.